### PR TITLE
Extensive refactoring and code cleanup

### PR DESCRIPTION
**Pull request by hisahi**
### Short description
See title.

This might need some time to merge and test properly and will likely cause large merge conflicts in any other PR started before this is merged. Two reviewers are recommended, but there is a lot to review.

### DoD
I have:
- [ ] added actual code.
- [x] produced clean code that passes ESLint.
- [x] code that actually does what it should.
- [ ] documented the code or added documentation to the wiki.
- [x] added or modified test(s).
- [x] test(s) that actually pass.

### CODE REVIEW

**Comment by gotonode** [comment]
Ping me for a review when needed. Do all tests pass?

**Comment by hisahi** [comment]
Tests are fine locally and on Travis.

**Comment by gotonode** [comment]
This looks much better than the previous code. Refactoring truly can pay off.

Should I add review comments on lines I think might contain issues?

Two reviewers are indeed a good thing.

**Comment by hisahi** [comment]
Please add comments where necessary.

File: labtool2.0/src/tests/__snapshots__/Courses.test.js.snap [hunk 1]
```diff
@@ -38,7 +38,7 @@ exports[`<Courses /> Courses Component should render correctly 1`] = `
               as="th"
               colSpan="1"
             >
-              Course id
+              Course ID
```
Comment (gotonode) [inline labtool2.0/src/tests/__snapshots__/Courses.test.js.snap]: I agree with the uppercase format.

File: labtool2.0/src/tests/__snapshots__/BrowseReviews.test.js.snap [hunk 2]
```diff
@@ -67,7 +67,7 @@ exports[`<BrowseReviews /> BrowseReviews Component should render correctly 1`] =
         <p
           className="noPrevious"
         >
-          First time to participate the course
+          Has not been on this course before
```
Comment (gotonode) [inline labtool2.0/src/tests/__snapshots__/BrowseReviews.test.js.snap]: "Has not taken part in/on this course before"?

Comment (hisahi) [reply labtool2.0/src/tests/__snapshots__/BrowseReviews.test.js.snap]: Seems better

**Comment by gotonode** [comment]
Many merge conflicts?

**Comment by hisahi** [comment]
#647 ought to be pulled first.

File: labtool2.0/src/components/LabtoolComment.js [hunk 3]
```diff
@@ -0,0 +1,62 @@
+import React from 'react'
+import { Button, Icon, Comment, Label } from 'semantic-ui-react'
+import ReactMarkdown from 'react-markdown'
+import { withRouter } from 'react-router'
+import { connect } from 'react-redux'
+import { trimDate } from '../util/format'
+
+export class LabtoolComment extends React.Component {
```
Comment (mjaakko) [inline labtool2.0/src/components/LabtoolComment.js]: This could be functional component

File: labtool2.0/src/components/StudentTable.js [hunk 4]
```diff
@@ -0,0 +1,442 @@
...
+import { Link } from 'react-router-dom'
+import { connect } from 'react-redux'
+import HorizontalScrollable from './HorizontalScrollable'
+import { getAllTags, tagStudent, unTagStudent } from '../services/tags'
+import { associateTeacherToStudent } from '../services/assistant'
+import { showAssistantDropdown, showTagDropdown, filterByTag, filterByAssistant, selectTeacher, selectTag } from '../reducers/coursePageLogicReducer'
+
+const { Fragment } = React
+
+export class StudentTable extends React.Component {
```
Comment (mjaakko) [inline labtool2.0/src/components/StudentTable.js]: This could be functional component

File: labtool2.0/src/components/LabtoolComment.js (hunk 3 above)
Comment (hisahi) [reply labtool2.0/src/components/LabtoolComment.js]: it is better to have these as class components for now with a functional component conversion being done in a later refactoring commit, converting _all_ components into being functional.

File: labtool2.0/src/components/StudentTable.js (hunk 4 above)
Comment (hisahi) [reply labtool2.0/src/components/StudentTable.js]: see comment in LabtoolComment.js: tl;dr, we should keep these as classes for now and convert all components into functional ones later on

**Pull request merged**